garden_ai/metadata.py

```diff
@@ -0,0 +1,224 @@
+            # Replace old cell source with new metadata
+            metadata_string = json.dumps(
+                {
+                    "NOTEBOOK_BASE_IMAGE_NAME": base_image_name,
+                    "NOTEBOOK_REQUIREMENTS": requirements_data,
+                },
+                indent=2,
+            )
+
+            cell["source"] = NOTEBOOK_METADATA_CELL_TEMPLATE.substitute(
+                metadata=metadata_string
+            )
+
+            # Add hidden metadata
+            cell["metadata"]["garden_metadata"][
+                "NOTEBOOK_BASE_IMAGE_URI"
+            ] = base_image_uri
+
+    # Write updated notebook data to file
+    nbformat.write(ntbk, notebook_path, version=nbformat.NO_CONVERT)
+
+
+def get_notebook_metadata(notebook_path: Path) -> dict:
+    notebook_metadata: dict = {}
+    notebook_metadata["notebook_image_name"] = None
+    notebook_metadata["notebook_requirements"] = None
+    notebook_metadata_cell_source = None
+    notebook_metadata_string = None
+
+    ntbk = _read_notebook(notebook_path)
+
+    # Find cell with 'garden_metadata_cell' tag and get source
+    for cell in ntbk.cells:
+        cell_tags = cell.get("metadata", {}).get("tags", [])
+        if "garden_metadata_cell" in cell_tags:
+            # Grab cell source and hidden metadata
+            notebook_metadata_cell_source = cell.get("source", None)
+            notebook_hidden_metadata = cell.get("metadata", {}).get(
+                "garden_metadata", {}
+            )
+            break
+
+    # Return empty notebook metadata dict if was unable to find cell source
+    if not notebook_metadata_cell_source:
+        typer.echo("Unable to find garden metadata cell.")
+        return notebook_metadata
+
+    # Grab the part of cell source with the metadata dict in it
+    clean_source = notebook_metadata_cell_source.replace("\n", "")
+    regex_get_metadata = r"^(.*?)\"\"\"(.*?)NOTEBOOK_METADATA(.*?)=(.*?)\"\"\"(.*?)$"
+
+    metadata_match = re.match(regex_get_metadata, clean_source)
+    if metadata_match and len(metadata_match.groups()) == 5:
+        notebook_metadata_string = metadata_match[4]
+
+    if notebook_metadata_string:
+        notebook_metadata_dict = json.loads(notebook_metadata_string.strip())
+        notebook_metadata["notebook_image_name"] = notebook_metadata_dict.get(
+            "NOTEBOOK_BASE_IMAGE_NAME", None
+        )
+        notebook_metadata["notebook_requirements"] = notebook_metadata_dict.get(
+            "NOTEBOOK_REQUIREMENTS", None
+        )
+        notebook_metadata["notebook_image_uri"] = notebook_hidden_metadata.get(
+            "NOTEBOOK_BASE_IMAGE_URI", None
+        )
+
+    return notebook_metadata
+
+
+def read_requirements_data(
+    requirements_path: Optional[Path],
+    notebook_path: Path,
+) -> Optional[dict]:
+    requirements_data: dict = {}
+
+    # Always use requirements from user provided requirements_path over any previously saved notebook requirements.
+    if requirements_path:
```

Comment:
Nit: this is a smell for me. It feels like the calling function should know whether it has a requirements path and then call the right helper as opposed to sending None down to the helper. This also means the helper knows about the precedence logic of user-provided requirements over the previously saved requirements. That feels like business logic that should live in the CLI app module.

I'm not 100% sure you should heed this nit. What you have might be better than alternatives. But throwing it out there and you can act on it if you think it would improve things